src/main/host/descriptor/socket/abstract_unix_ns.rs

```diff
@@ -0,0 +1,273 @@
+                    name = Some(temp_name.to_vec());
+                    break;
+                }
+            }
+        }
+
+        let name = match name {
+            Some(x) => x,
+            // every valid name has been taken
+            None => return Err(BindError::NoNamesAvailable),
+        };
+
+        let name_copy = name.clone();
+
+        // when the socket closes, remove this entry from the namespace
+        let handle = Self::on_socket_close(
+            Arc::downgrade(&ns_arc),
+            &mut socket.borrow_mut(),
+            move |ns| {
+                assert!(ns.unbind(&name_copy).is_ok());
+            },
+        );
+
+        if let std::collections::hash_map::Entry::Vacant(entry) = ns.address_map.entry(name) {
+            entry.insert((Arc::downgrade(socket), handle));
+        } else {
+            unreachable!();
+        }
+
+        Ok(())
+    }
+
+    pub fn unbind(&mut self, name: &Vec<u8>) -> Result<(), BindError> {
+        // remove the namespace entry which includes the handle, so the event listener will
+        // automatically be removed from the socket
+        if self.address_map.remove(name).is_none() {
+            // didn't exist in the address map
+            return Err(BindError::NameNotFound);
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum BindError {
+    /// The name is already in use.
+    NameInUse,
+    /// Names in the ephemeral name range are all in use.
+    NoNamesAvailable,
+    /// The name was not found in the address map.
+    NameNotFound,
+}
+
+impl std::error::Error for BindError {}
+
+impl std::fmt::Display for BindError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::NameInUse => write!(f, "Name is already in use"),
+            Self::NoNamesAvailable => {
+                write!(f, "Names in the ephemeral name range are all in use")
+            }
+            Self::NameNotFound => write!(f, "Name was not found in the address map"),
+        }
+    }
+}
+
+/// The characters that are valid in auto-generated names.
+const CHARSET: &[u8] = b"abcdef0123456789";
+const NAME_LEN: usize = 5;
+
+/// Choose a random name of length `L`.
+fn random_name<const L: usize>(mut rng: impl rand::Rng) -> [u8; L] {
+    let mut name = [0u8; L];
+
+    // set each character of the name
+    for x in 0..L {
+        name[x] = *CHARSET.choose(&mut rng).unwrap();
+    }
+
+    name
+}
+
+/// Get a name in the set of all valid names. This would be better implemented as a generator when
+/// generators become stable.
+fn incremental_name<const L: usize>(index: usize) -> [u8; L] {
+    // there are a limited number of valid names
+    assert!(index < CHARSET.len().pow(L as u32));
+
+    let mut name = [0u8; L];
+
+    // set each character of the name
+    for x in 0..L {
+        let charset_index = (index / CHARSET.len().pow(x as u32)) % CHARSET.len();
```

Comment:
This is a bit tough to follow. Maybe this can be broken up into multiple statements and/or commented a bit to help explain how this works?

Reply:
I think I improved this, it's essentially just changing the index from base 10 to the digits as base `CHARSET.len()` (16).